apps/server/lib/lexical/server/provider/handlers/rename.ex

```diff
@@ -0,0 +1,80 @@
+defmodule Lexical.Server.Provider.Handlers.Rename do
+  alias Lexical.Ast
+  alias Lexical.Document
+  alias Lexical.Protocol.Requests.Rename
+  alias Lexical.Protocol.Responses
+  alias Lexical.Protocol.Types.RenameFile
+  alias Lexical.Protocol.Types.TextDocument
+  alias Lexical.Protocol.Types.Workspace
+  alias Lexical.RemoteControl.Api
+  alias Lexical.Server.Provider.Env
+  require Logger
+
+  def handle(%Rename{} = request, %Env{} = env) do
+    case Document.Store.fetch(request.document.uri, :analysis) do
+      {:ok, _document, %Ast.Analysis{valid?: true} = analysis} ->
+        rename(
+          env.project,
+          analysis,
+          request.position,
+          request.new_name,
+          request.id,
+          env.client_name
+        )
+
+      _ ->
+        {:reply,
+         Responses.Rename.error(request.id, :request_failed, "document can not be analyzed")}
+    end
+  end
+
+  defp rename(project, analysis, position, new_name, id, client_name) do
+    case Api.rename(project, analysis, position, new_name, client_name) do
+      {:ok, []} ->
+        {:reply, nil}
+
+      {:ok, results} ->
+        text_document_edits =
+          Enum.map(results, fn %Document.Changes{edits: edits, document: document} ->
+            new_text_document_edit(document.uri, edits)
+          end)
+
+        rename_files =
+          results
+          |> Stream.map(& &1.rename_file)
+          |> Stream.reject(&(&1 == nil))
+          |> Enum.map(&new_rename_file/1)
+
+        workspace_edit = Workspace.Edit.new(document_changes: text_document_edits ++ rename_files)
+
+        {:reply, Responses.Rename.new(id, workspace_edit)}
+
+      {:error, {:unsupported_entity, entity}} ->
+        Logger.info("Unrenameable entity: #{inspect(entity)}")
+        {:reply, nil}
+
+      {:error, reason} ->
+        {:reply, Responses.Rename.error(id, :request_failed, inspect(reason))}
+    end
+  end
+
+  defp new_text_document_edit(uri, edits) do
+    # NOTE: the `0` here is for use in VSCode.
+    # Before May 4, 2024, if this `0` number is not set,
+    # the rename function in VSCode will not work,
```

Comment:
you sure that the reason is that you're not setting the correct version of opened documents?

Reply:
You are right, simply using `document.version` here will suffice, fixed